File: python/src/ttio/isolation_window.py
"""``IsolationWindow`` — precursor isolation window for MS/MS scans."""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class IsolationWindow:
    """Precursor isolation window, expressed as a target m/z with
    asymmetric lower/upper offsets in Th (Da).

    The instrument-reported window spans
    ``[target_mz - lower_offset, target_mz + upper_offset]``. Offsets are
    non-negative by convention; the lower and upper may differ when the
    quadrupole is offset from the monoisotopic m/z (common in DIA).

    Parameters
    ----------
    target_mz : float
        Center of the isolation window in Th (Da).
    lower_offset : float
        Distance from ``target_mz`` to the lower bound, in Th. Non-negative.
    upper_offset : float
        Distance from ``target_mz`` to the upper bound, in Th. Non-negative.

    Notes
    -----
    API status: Stable (v1.1, M74).

    Cross-language equivalents
    --------------------------
    Objective-C: ``TTIOIsolationWindow`` · Java:
    ``com.dtwthalion.ttio.IsolationWindow``.
    """

    target_mz: float
    lower_offset: float
    upper_offset: float

    @property
    def lower_bound(self) -> float:
        """Lower m/z bound ``target_mz - lower_offset``."""
        return self.target_mz - self.lower_offset

    @property
    def upper_bound(self) -> float:
        """Upper m/z bound ``target_mz + upper_offset``."""
        return self.target_mz + self.upper_offset

    @property
    def width(self) -> float:
        """Total isolation width ``lower_offset + upper_offset``."""
        return self.lower_offset + self.upper_offset

File: java/src/main/java/global/thalion/ttio/IsolationWindow.java
/*
 * TTI-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio;

/**
 * Precursor isolation window for MS/MS scans, expressed as a target
 * m/z with asymmetric lower/upper offsets in Th (Da).
 *
 * <p>The instrument-reported window spans
 * {@code [targetMz - lowerOffset, targetMz + upperOffset]}. Offsets are
 * non-negative by convention; the lower and upper may differ when the
 * quadrupole is offset from the monoisotopic m/z (common in DIA).</p>
 *
 * <p><b>API status:</b> Stable (v1.1, M74).</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOIsolationWindow}, Python
 * {@code ttio.isolation_window.IsolationWindow}.</p>
 *
 *
 */
public record IsolationWindow(double targetMz, double lowerOffset,
                              double upperOffset) {

    /** @return lower m/z bound {@code targetMz - lowerOffset}. */
    public double lowerBound() {
        return targetMz - lowerOffset;
    }

    /** @return upper m/z bound {@code targetMz + upperOffset}. */
    public double upperBound() {
        return targetMz + upperOffset;
    }

    /** @return total isolation width {@code lowerOffset + upperOffset}. */
    public double width() {
        return lowerOffset + upperOffset;
    }
}
